executor/cte.go

```diff
@@ -0,0 +1,434 @@
+			return err
+		}
+	}
+	close(e.iterInTbl.GetBegCh())
+
+	// Clear data in iterOutTbl.
+	return e.iterOutTbl.Reopen()
+}
+
+func (e *CTEExec) reset() {
+	e.curIter = 0
+	e.chkIdx = 0
+	e.iterInHashTbl = nil
+	e.resHashTbl = nil
+}
+
+func (e *CTEExec) reopenTbls() (err error) {
+	if err := e.resTbl.Reopen(); err != nil {
+		return err
+	}
+	return e.iterInTbl.Reopen()
+}
+
+func setupCTEStorageTracker(tbl cteutil.Storage, ctx sessionctx.Context) {
+	memTracker := tbl.GetMemTracker()
+	memTracker.SetLabel(memory.LabelForCTEStorage)
+	memTracker.AttachTo(ctx.GetSessionVars().StmtCtx.MemTracker)
+
+	diskTracker := tbl.GetDiskTracker()
+	diskTracker.SetLabel(memory.LabelForCTEStorage)
+	diskTracker.AttachTo(ctx.GetSessionVars().StmtCtx.DiskTracker)
+
+	if config.GetGlobalConfig().OOMUseTmpStorage {
+		actionSpill := tbl.ActionSpill()
+		ctx.GetSessionVars().StmtCtx.MemTracker.FallbackOldAndSetNewAction(actionSpill)
+	}
+}
+
+func (e *CTEExec) tryFilterAndAdd(chk *chunk.Chunk,
+	storage cteutil.Storage,
+	hashtbl baseHashTable) (res *chunk.Chunk, err error) {
+	if e.isDistinct {
+		if chk, err = e.filterAndAddHashTable(chk, storage, hashtbl); err != nil {
+			return nil, err
+		}
+	}
+	if chk.NumRows() == 0 {
+		return chk, nil
+	}
+	return chk, storage.Add(chk)
+}
+
+func (e *CTEExec) filterAndAddHashTable(chk *chunk.Chunk,
+	storage cteutil.Storage,
+	hashtbl baseHashTable) (finalChkNoDup *chunk.Chunk, err error) {
+	rows := chk.NumRows()
+	if rows == 0 {
+		return chk, nil
+	}
+
+	buf := make([]byte, 1)
+	isNull := make([]bool, rows)
+	hasher := make([]hash.Hash64, rows)
+	for i := 0; i < rows; i++ {
+		// New64() just returns a int64 constant,
+		// but we can avoid calling it every time this func is called.
+		hasher[i] = fnv.New64()
+	}
+
+	tps := e.base().retFieldTypes
+	for i := 0; i < chk.NumCols(); i++ {
+		err = codec.HashChunkColumns(e.ctx.GetSessionVars().StmtCtx, hasher, chk, tps[i], i, buf, isNull)
+		if err != nil {
+			return nil, errors.Trace(err)
+		}
+	}
+	tmpChkNoDup := chunk.NewChunkWithCapacity(tps, chk.Capacity())
+	chkHashTbl := newConcurrentMapHashTable()
+	idxForOriRows := make([]int, 0, chk.NumRows())
+
+	// filter rows duplicated in cur chk
+	for i := 0; i < rows; i++ {
+		key := hasher[i].Sum64()
+		row := chk.GetRow(i)
+
+		hasDup, err := e.checkHasDup(key, row, chk, storage, chkHashTbl)
+		if err != nil {
+			return nil, err
+		}
+		if hasDup {
+			continue
+		}
+
+		tmpChkNoDup.AppendRow(row)
+
+		rowPtr := chunk.RowPtr{ChkIdx: uint32(0), RowIdx: uint32(i)}
+		chkHashTbl.Put(key, rowPtr)
+		idxForOriRows = append(idxForOriRows, i)
+	}
+
+	// filter rows duplicated in RowContainer
+	chkIdx := storage.NumChunks()
+	finalChkNoDup = chunk.NewChunkWithCapacity(tps, chk.Capacity())
+	for i := 0; i < tmpChkNoDup.NumRows(); i++ {
+		key := hasher[idxForOriRows[i]].Sum64()
+		row := tmpChkNoDup.GetRow(i)
+
+		hasDup, err := e.checkHasDup(key, row, nil, storage, hashtbl)
```

Comment:
`checkHasDup` will encode encode `row` again in EqualChunkRow, this is an unnecessary CPU overhead. Can we refine this?